File: ai-service/utils/video_source.py
"""Validate public source identifiers without accessing files or the network."""
import re
from urllib.parse import parse_qs, urlsplit

HOSTS = {"youtube.com", "www.youtube.com", "m.youtube.com", "music.youtube.com", "youtu.be"}
INVALID_URL = "Provide an HTTPS YouTube video URL (watch, youtu.be, shorts, embed, or live)."


def normalize_youtube_url(value: str) -> str:
    if not isinstance(value, str) or len(value) > 2048:
        raise ValueError(INVALID_URL)
    value = value.strip()
    if not value.lower().startswith("https://") or re.search(r"[\s\\\x00-\x1f\x7f]", value):
        raise ValueError(INVALID_URL)
    try:
        url = urlsplit(value)
        authority = url.netloc.lower()
        if authority.removesuffix(":443") not in HOSTS or url.hostname not in HOSTS:
            raise ValueError(INVALID_URL)
        if "%" in url.path or ".." in url.path:
            raise ValueError(INVALID_URL)
        query = parse_qs(url.query, keep_blank_values=True)
        if any(key.lower() == "list" for key in query) or url.path == "/playlist":
            raise ValueError("Playlists are not supported. Submit a single video without the list parameter.")
        video_id = None
        if url.hostname == "youtu.be":
            match = re.fullmatch(r"/([A-Za-z0-9_-]{11})/?", url.path)
            video_id = match[1] if match else None
        elif url.path == "/watch":
            ids = query.get("v", [])
            video_id = ids[0] if len(ids) == 1 else None
        else:
            match = re.fullmatch(r"/(?:shorts|embed|live)/([A-Za-z0-9_-]{11})/?", url.path)
            video_id = match[1] if match else None
        if not video_id or not re.fullmatch(r"[A-Za-z0-9_-]{11}", video_id):
            raise ValueError(INVALID_URL)
    except (TypeError, AttributeError) as exc:
        raise ValueError(INVALID_URL) from exc
    return f"https://www.youtube.com/watch?v={video_id}"


def validate_video_source(youtube_url, file):
    has_url = youtube_url is not None
    if has_url == (file is not None):
        raise ValueError("Provide exactly one source: youtube_url or a file upload.")
    return ("youtube", normalize_youtube_url(youtube_url)) if has_url else ("upload", None)

File: server/src/utils/videoSource.js
import { ApiError } from "./ApiError.js";

const HOSTS = new Set(["youtube.com", "www.youtube.com", "m.youtube.com", "music.youtube.com", "youtu.be"]);
const VIDEO_ID = /^[A-Za-z0-9_-]{11}$/;
const INVALID_URL = "Provide an HTTPS YouTube video URL (watch, youtu.be, shorts, embed, or live).";

export function normalizeYouTubeUrl(value) {
  if (typeof value !== "string" || value.length > 2048) throw new ApiError(400, INVALID_URL);
  const input = value.trim();
  if (!/^https:\/\//i.test(input) || /[\s\\\x00-\x1f\x7f]/.test(input)) {
    throw new ApiError(400, INVALID_URL);
  }
  let url;
  try { url = new URL(input); } catch { throw new ApiError(400, INVALID_URL); }
  const authority = input.slice(8).split(/[/?#]/, 1)[0].toLowerCase();
  // Checking the raw authority also rejects encoded hosts, userinfo and odd ports.
  if (!HOSTS.has(authority.replace(/:443$/, "")) || !HOSTS.has(url.hostname)) {
    throw new ApiError(400, INVALID_URL);
  }
  const rawPath = input.slice(8 + authority.length).split(/[?#]/, 1)[0];
  if (rawPath.includes("%") || rawPath.includes("..")) throw new ApiError(400, INVALID_URL);
  if ([...url.searchParams.keys()].some((key) => key.toLowerCase() === "list") || url.pathname === "/playlist") {
    throw new ApiError(400, "Playlists are not supported. Submit a single video without the list parameter.");
  }
  let id;
  if (url.hostname === "youtu.be") {
    id = /^\/([A-Za-z0-9_-]{11})\/?$/.exec(url.pathname)?.[1];
  } else if (url.pathname === "/watch") {
    const ids = url.searchParams.getAll("v");
    if (ids.length === 1) id = ids[0];
  } else {
    id = /^\/(?:shorts|embed|live)\/([A-Za-z0-9_-]{11})\/?$/.exec(url.pathname)?.[1];
  }
  if (!id || !VIDEO_ID.test(id)) throw new ApiError(400, INVALID_URL);
  return `https://www.youtube.com/watch?v=${id}`;
}

export function validateVideoSource(body = {}, file) {
  if (!body || typeof body !== "object" || Array.isArray(body)) {
    throw new ApiError(400, "Provide exactly one source: youtubeUrl or a file upload.");
  }
  const hasUrl = Object.prototype.hasOwnProperty.call(body, "youtubeUrl");
  if (hasUrl === Boolean(file)) {
    throw new ApiError(400, "Provide exactly one source: youtubeUrl or a file upload.");
  }
  return hasUrl
    ? { type: "youtube", youtubeUrl: normalizeYouTubeUrl(body.youtubeUrl) }
    : { type: "upload", file };
}
